src/models/GitError.js

```diff
@@ -0,0 +1,147 @@
+  NoHeadCommitError: `NoHeadCommitError`,
+  CommitNotFetchedError: `CommitNotFetchedError`,
+  ObjectTypeUnknownFail: `ObjectTypeUnknownFail`,
+  ObjectTypeAssertionFail: `ObjectTypeAssertionFail`,
+  ObjectTypeAssertionInTreeFail: `ObjectTypeAssertionInTreeFail`,
+  ObjectTypeAssertionInRefFail: `ObjectTypeAssertionInRefFail`,
+  ObjectTypeAssertionInPathFail: `ObjectTypeAssertionInPathFail`,
+  MissingAuthorError: `MissingAuthorError`,
+  GitRootNotFoundError: `GitRootNotFoundError`,
+  UnparseableServerResponseFail: `UnparseableServerResponseFail`,
+  InvalidDepthParameterError: `InvalidDepthParameterError`,
+  RemoteDoesNotSupportShallowFail: `RemoteDoesNotSupportShallowFail`,
+  RemoteDoesNotSupportDeepenSinceFail: `RemoteDoesNotSupportDeepenSinceFail`,
+  RemoteDoesNotSupportDeepenNotFail: `RemoteDoesNotSupportDeepenNotFail`,
+  RemoteDoesNotSupportDeepenRelativeFail: `RemoteDoesNotSupportDeepenRelativeFail`,
+  CorruptShallowOidFail: `CorruptShallowOidFail`,
+  FastForwardFail: `FastForwardFail`,
+  MergeNotSupportedFail: `MergeNotSupportedFail`,
+  DirectorySeparatorsError: `DirectorySeparatorsError`,
+  ResolveTreeError: `ResolveTreeError`,
+  DirectoryIsAFileError: `DirectoryIsAFileError`,
+  TreeOrBlobNotFoundError: `TreeOrBlobNotFoundError`,
+  NotImplementedFail: `NotImplementedFail`,
+  ReadObjectFail: `ReadObjectFail`,
+  ReadShallowObjectFail: `ReadShallowObjectFail`,
+  NotAnOidFail: `NotAnOidFail`,
+  NoRefspecConfiguredError: `NoRefspecConfiguredError`,
+  ResolveRefError: `ResolveRefError`,
+  ExpandRefError: `ExpandRefError`,
+  EmptyServerResponseFail: `EmptyServerResponseFail`,
+  AssertServerResponseFail: `AssertServerResponseFail`,
+  HTTPError: `HTTPError`,
+  RemoteUrlParseError: `RemoteUrlParseError`,
+  UnknownTransportError: `UnknownTransportError`,
+  AcquireLockFileFail: `AcquireLockFileFail`,
+  DoubleReleaseLockFileFail: `DoubleReleaseLockFileFail`,
+  InternalFail: `InternalFail`,
+  UnknownOauth2Format: `UnknownOauth2Format`,
+  MissingPasswordTokenError: `MissingPasswordTokenError`,
+  MissingUsernameError: `MissingUsernameError`,
+  MixPasswordTokenError: `MixPasswordTokenError`,
+  MixUsernamePasswordTokenError: `MixUsernamePasswordTokenError`,
+  MissingTokenError: `MissingTokenError`,
+  MixUsernameOauth2formatMissingTokenError: `MixUsernameOauth2formatMissingTokenError`,
+  MixPasswordOauth2formatMissingTokenError: `MixPasswordOauth2formatMissingTokenError`,
+  MixUsernamePasswordOauth2formatMissingTokenError: `MixUsernamePasswordOauth2formatMissingTokenError`,
+  MixUsernameOauth2formatTokenError: `MixUsernameOauth2formatTokenError`,
+  MixPasswordOauth2formatTokenError: `MixPasswordOauth2formatTokenError`,
+  MixUsernamePasswordOauth2formatTokenError: `MixUsernamePasswordOauth2formatTokenError`
+}
+
+export class GitError extends Error {
+  constructor (code, data) {
+    super()
+    this.name = code
+    this.code = code
+    this.data = data
+    this.message = messages[code](data || {})
```

Comment:
Maybe this should be 
```javascript
this.message = data.message || messages[code](data || {})
delete data.message; // but first data need to be cloned to don't delete user field
this.data = data
```

Reply:
Ooooh. Oooh. Is it _too_ clever? I like it.

Reply:
I think it's _too_ clever. I think I'll just make a second constructor called `GitInternalError`.

Reply:
On third thought... I don't see why I should worry about this at all. I think it's fine the way it is.